File: tests/test_pauli.py
from sparse_sim.cython.core import *


def test_pauli_string_initialization_scaling_freeing():
    paulis0 = ['I', 'X', 'I', 'X']
    pString0 = PauliString(4, 1 + 0j, paulis0)

    paulis1 = ['I', 'X', 'I', 'X']
    pString1 = PauliString(4, 1 + 1j, paulis1)

    print(pString0)

    pString2 = (2 + 0j) * pString1
    print(f"2 * {pString1} = {pString2}")

    pString3 = pString2.adjoint()
    print(f"Adjoint of {pString2} = {pString3}")


def test_pauli_string_multiplication():
    paulis0 = ["I", "Y", "I", "X"]
    pString0 = PauliString(4, 1 + 0j, paulis0)

    paulis1 = ["I", "X", "I", "I"]
    pString1 = PauliString(4, 1j, paulis1)

    pString2 = pString0 * pString1

    print(f"{pString0} * {pString1} = {pString2}")


def test_pauli_sum_initialization_scaling_freeing():
    paulis0 = ['I', 'Y', 'I', 'X']
    pString0 = PauliString(4, 1 + 0j, paulis0)

    paulis1 = ['I', 'X', 'I', 'I']
    pString1 = PauliString(4, 1j, paulis1)

    pSum0 = PauliSum(4)
    pSum0.append_pauli_string(pString0)
    pSum0.append_pauli_string(pString1)

    print(f"{pSum0} contains {pSum0.p} pStrings")
    pSum1 = (2 + 0j) * pSum0
    print(f"2 * {pSum0} = {pSum1}")


def test_pauli_string_to_matrix():
    paulis0 = ['I', 'X']
    pString0 = PauliString(2, 1 + 0j, paulis0)

    paulis1 = ['Y', 'X']
    pString1 = PauliString(2, 0.5j, paulis1)

    print(f"{pString0} ->\n {pString0.matrix}")
    print(f"{pString1} ->\n {pString1.matrix}")


def test_pauli_sum_multiplication():
    paulis0 = ['I', 'Y', 'I', 'X']
    pString0 = PauliString(4, 1 + 0j, paulis0)

    paulis1 = ['I', 'X', 'I', 'I']
    pString1 = PauliString(4, 1j, paulis1)

    pSum0 = PauliSum(4)
    pSum0.append_pauli_string(pString0)
    pSum0.append_pauli_string(pString1)

    paulis2 = ['I', 'Y', 'I', 'X']
    pString2 = PauliString(4, 1 + 0j, paulis2)

    paulis3 = ['I', 'X', 'I', 'I']
    pString3 = PauliString(4, 1j, paulis3)

    pSum1 = PauliSum(4)
    pSum1.append_pauli_string(pString2)
    pSum1.append_pauli_string(pString3)

    pSum2 = pSum0 * pSum1

    print(f"({pSum0}) * ({pSum1}) = {pSum2}")


def test_pauli_sum_addition():
    paulis0 = ['I', 'Y', 'I', 'X']
    pString0 = PauliString(4, 1 + 0j, paulis0)

    paulis1 = ['I', 'X', 'I', 'I']
    pString1 = PauliString(4, 1j, paulis1)

    pSum0 = PauliSum(4)
    pSum0.append_pauli_string(pString0)
    pSum0.append_pauli_string(pString1)

    paulis2 = ['I', 'Y', 'I', 'X']
    pString2 = PauliString(4, 1 + 0j, paulis2)

    paulis3 = ['I', 'X', 'I', 'I']
    pString3 = PauliString(4, -1j, paulis3)

    pSum1 = PauliSum(4)
    pSum1.append_pauli_string(pString2)
    pSum1.append_pauli_string(pString3)

    pSum2 = pSum0 + pSum1

    print(f"({pSum0}) + ({pSum1}) = {pSum2}")

    pSum3 = -1 * pSum2
    pSum4 = pSum2 + pSum3
    print(f"{pSum2} + {pSum3} = {pSum4}")


def test_pauli_sum_get_pauli_strings():
    paulis0 = ['Z', 'Y', 'I', 'X']
    paulis2 = ['Y', 'Y', 'I', 'X']
    paulis1 = ['I', 'X', 'Y', 'I']
    paulis3 = ['I', 'X', 'I', 'I']

    pString0 = PauliString(4, 1 + 0j, paulis0)
    print(paulis0)
    pString1 = PauliString(4, 1j, paulis1)

    print(paulis1)
    pString2 = PauliString(4, 1 + 0j, paulis2)

    print(paulis2)
    pString3 = PauliString(4, -1j, paulis3)

    print(paulis3)

    pSum0 = PauliSum(4)
    pSum0.append_pauli_string(pString0)
    pSum0.append_pauli_string(pString1)
    pSum0.append_pauli_string(pString2)
    pSum0.append_pauli_string(pString3)
    pauli_strings = pSum0.get_pauli_strings()
    if not pauli_strings:
        print("Error: Failed to retrieve Pauli strings.")
        return

    print(f"{pSum0} contains {pSum0.p} Pauli strings:")
    for i, p_str in enumerate(pauli_strings):
        print(f"Pauli string {i}: {p_str}")


def test_pauli_sum_to_matrix():
    paulis0 = ['I', 'X']
    pString0 = PauliString(2, 1 + 0j, paulis0)

    paulis1 = ['Y', 'X']
    pString1 = PauliString(2, 0.5j, paulis1)

    pSum0 = pString0 + pString1

    print(f"{pString0} ->\n {pString0.matrix}")
    print(f"{pString1} ->\n {pString1.matrix}")
    print(f"{pSum0} ->\n {pSum0.matrix}")


def main():
    print("\nTesting PauliString initialization, scaling, and freeing")
    test_pauli_string_initialization_scaling_freeing()

    print("\nTesting PauliString to Matrix")
    test_pauli_string_to_matrix()

    print("\nTesting PauliString Multiplication")
    test_pauli_string_multiplication()

    print("\nTesting PauliSum initialization, scaling, and freeing")
    test_pauli_sum_initialization_scaling_freeing()

    print("\nTesting PauliSum Multiplication")
    test_pauli_sum_multiplication()

    print("\nTesting PauliSum addition")
    test_pauli_sum_addition()

    print("\nTesting PauliSum get_pauli_strings")
    test_pauli_sum_get_pauli_strings()

    print("\nTesting PauliSum to Matrix")
    test_pauli_sum_to_matrix()


if __name__ == "__main__":
    main()

File: src/sparse_sim/src/test_pauli.c
#include <complex.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "pauli.h"

void test_pauli_string_initialization_scaling_freeing(void);
void test_pauli_string_multiplication(void);
void test_pauli_sum_initialization_scaling_freeing(void);
void test_pauli_sum_adjoint(void);
void test_pauli_sum_multiplication(void);
void test_pauli_sum_addition(void);
void test_pauli_encoding(unsigned int N);
void test_pauli_sum_get_pauli_strings(void);

void test_pauli_string_initialization_scaling_freeing(void) {
  unsigned int paulis0[] = {0, 1, 0, 1};
  char paulis1[] = {'I', 'X', 'I', 'X'};
  PauliStringC *pString0;
  PauliStringC *pString1;
  PauliStringC *pString2;
  PauliStringC *pString3;
  char *pString0_str;
  char *pString1_str;
  char *pString2_str;
  char *pString3_str;

  pString0 = pauli_string_init_as_ints_c(4, (double complex)1.0, paulis0);
  pString1 = pauli_string_init_as_chars_c(4, (double complex)I, paulis1);
  pString0_str = pauli_string_to_string_c(pString0);
  printf("%s with encoding %llu\n", pString0_str, pString0->encoding);

  pString1_str = pauli_string_to_string_c(pString1);

  pString2 = pauli_string_scalar_multiplication_c(pString1, 2);
  pString2_str = pauli_string_to_string_c(pString2);
  printf("2 * %s = %s\n", pString1_str, pString2_str);

  pString3 = pauli_string_adjoint_c(pString2);
  pString3_str = pauli_string_to_string_c(pString3);
  printf("Adjoint of %s = %s\n", pString2_str, pString3_str);

  free(pString0_str);
  free(pString1_str);
  free(pString2_str);
  free(pString3_str);

  free_pauli_string_c(pString0);
  free_pauli_string_c(pString1);
  free_pauli_string_c(pString2);
  free_pauli_string_c(pString3);
}

void test_pauli_string_multiplication(void) {
  unsigned int paulis0[] = {0, 2, 0, 1};
  char paulis1[] = {'I', 'X', 'I', 'I'};
  char *pString0_str;
  char *pString1_str;
  char *pString2_str;
  PauliStringC *pString0;
  PauliStringC *pString1;
  PauliStringC *pString2;

  pString0 = pauli_string_init_as_ints_c(4, (double complex)1, paulis0);
  pString1 = pauli_string_init_as_chars_c(4, (double complex)I, paulis1);
  pString2 = pauli_string_multiplication_c(pString0, pString1);

  pString0_str = pauli_string_to_string_c(pString0);
  pString1_str = pauli_string_to_string_c(pString1);
  pString2_str = pauli_string_to_string_c(pString2);

  printf("%s * %s = %s\n", pString0_str, pString1_str, pString2_str);

  free(pString0_str);
  free(pString1_str);
  free(pString2_str);
  free_pauli_string_c(pString0);
  free_pauli_string_c(pString1);
  free_pauli_string_c(pString2);
}

void test_pauli_sum_initialization_scaling_freeing(void) {
  unsigned int paulis0[] = {0, 2, 0, 1};
  char paulis1[] = {'I', 'X', 'I', 'I'};
  char *pSum_str;
  PauliStringC *pString0;
  PauliStringC *pString1;
  PauliSumC *pSum;

  pString0 = pauli_string_init_as_ints_c(4, (double complex)1, paulis0);
  pString1 = pauli_string_init_as_chars_c(4, (double complex)I, paulis1);

  pSum = pauli_sum_init_c(4);
  pauli_sum_append_pauli_string_c(pSum, pString0);
  pauli_sum_append_pauli_string_c(pSum, pString1);

  pSum_str = pauli_sum_to_string_c(pSum);
  printf("%s contains %u pStrings of 2\n", pSum_str, pSum->p);

  free(pSum_str);
  free_pauli_sum_c(pSum);
}

void test_pauli_sum_adjoint(void) {
  unsigned int paulis0[] = {0, 2, 0, 1};
  unsigned int paulis2[] = {0, 2, 1, 1};
  char paulis1[] = {'I', 'X', 'X', 'I'};
  char paulis3[] = {'I', 'X', 'I', 'I'};

  char *pSum0_str;
  char *pSum1_str;

  PauliStringC *pString0;
  PauliStringC *pString1;
  PauliStringC *pString2;
  PauliStringC *pString3;

  PauliSumC *pSum0;
  PauliSumC *pSum1;

  pString0 = pauli_string_init_as_ints_c(4, (double complex)1, paulis0);
  pString1 = pauli_string_init_as_chars_c(4, (double complex)I, paulis1);
  pString2 = pauli_string_init_as_ints_c(4, (double complex)1, paulis2);
  pString3 = pauli_string_init_as_chars_c(4, (double complex) - I, paulis3);

  pSum0 = pauli_sum_init_c(4);
  pauli_sum_append_pauli_string_c(pSum0, pString0);
  pauli_sum_append_pauli_string_c(pSum0, pString1);
  pauli_sum_append_pauli_string_c(pSum0, pString2);
  pauli_sum_append_pauli_string_c(pSum0, pString3);

  pSum1 = pauli_sum_adjoint_c(pSum0);

  pSum0_str = pauli_sum_to_string_c(pSum0);
  pSum1_str = pauli_sum_to_string_c(pSum1);

  printf("(%s)^dagger = (%s)\n", pSum0_str, pSum1_str);

  free(pSum0_str);
  free(pSum1_str);

  free_pauli_sum_c(pSum0);
  free_pauli_sum_c(pSum1);
}

void test_pauli_sum_multiplication(void) {
  unsigned int paulis0[] = {0, 2, 0, 1};
  unsigned int paulis2[] = {0, 2, 0, 1};
  char paulis1[] = {'I', 'X', 'I', 'I'};
  char paulis3[] = {'I', 'X', 'I', 'I'};

  char *pSum0_str;
  char *pSum1_str;
  char *pSum2_str;

  PauliStringC *pString0;
  PauliStringC *pString1;
  PauliStringC *pString2;
  PauliStringC *pString3;

  PauliSumC *pSum0;
  PauliSumC *pSum1;
  PauliSumC *pSum2;

  pString0 = pauli_string_init_as_ints_c(4, (double complex)1, paulis0);
  pString1 = pauli_string_init_as_chars_c(4, (double complex)I, paulis1);
  pString2 = pauli_string_init_as_ints_c(4, (double complex)1, paulis2);
  pString3 = pauli_string_init_as_chars_c(4, (double complex)I, paulis3);

  pSum0 = pauli_sum_init_c(4);
  pauli_sum_append_pauli_string_c(pSum0, pString0);
  pauli_sum_append_pauli_string_c(pSum0, pString1);

  pSum1 = pauli_sum_init_c(4);
  pauli_sum_append_pauli_string_c(pSum1, pString2);
  pauli_sum_append_pauli_string_c(pSum1, pString3);

  pSum2 = pauli_sum_multiplication_c(pSum0, pSum1);

  pSum0_str = pauli_sum_to_string_c(pSum0);
  pSum1_str = pauli_sum_to_string_c(pSum1);
  pSum2_str = pauli_sum_to_string_c(pSum2);

  printf("(%s) * (%s) = (%s)\n", pSum0_str, pSum1_str, pSum2_str);

  free(pSum0_str);
  free(pSum1_str);
  free(pSum2_str);

  free_pauli_sum_c(pSum0);
  free_pauli_sum_c(pSum1);
  free_pauli_sum_c(pSum2);
}

void test_pauli_sum_addition(void) {
  unsigned int paulis0[] = {0, 2, 0, 1};
  unsigned int paulis2[] = {0, 2, 0, 1};
  char paulis1[] = {'I', 'X', 'I', 'I'};
  char paulis3[] = {'I', 'X', 'I', 'I'};

  char *pSum0_str;
  char *pSum1_str;
  char *pSum2_str;

  PauliStringC *pString0;
  PauliStringC *pString1;
  PauliStringC *pString2;
  PauliStringC *pString3;

  PauliSumC *pSum0;
  PauliSumC *pSum1;
  PauliSumC *pSum2;

  pString0 = pauli_string_init_as_ints_c(4, (double complex)1, paulis0);
  pString1 = pauli_string_init_as_chars_c(4, (double complex)I, paulis1);
  pString2 = pauli_string_init_as_ints_c(4, (double complex)1, paulis2);
  pString3 = pauli_string_init_as_chars_c(4, (double complex) - I, paulis3);

  pSum0 = pauli_sum_init_c(4);
  pauli_sum_append_pauli_string_c(pSum0, pString0);
  pauli_sum_append_pauli_string_c(pSum0, pString1);

  pSum1 = pauli_sum_init_c(4);
  pauli_sum_append_pauli_string_c(pSum1, pString2);
  pauli_sum_append_pauli_string_c(pSum1, pString3);

  pSum2 = pauli_sum_addition_c(pSum0, pSum1);

  pSum0_str = pauli_sum_to_string_c(pSum0);
  pSum1_str = pauli_sum_to_string_c(pSum1);
  pSum2_str = pauli_sum_to_string_c(pSum2);

  printf("(%s) + (%s) = (%s)\n", pSum0_str, pSum1_str, pSum2_str);

  free(pSum0_str);
  free(pSum1_str);
  free(pSum2_str);

  free_pauli_sum_c(pSum0);
  free_pauli_sum_c(pSum1);
  free_pauli_sum_c(pSum2);
}

void test_pauli_encoding(unsigned int N) {
  uint64_t total_strings = 1ULL << (2 * N);
  char pauli_chars[] = {'I', 'X', 'Y', 'Z'};

  printf("Testing %llu possible Pauli strings for %u qubits...\n",
         total_strings, N);

  char *paulis = (char *)malloc(N * sizeof(char));
  if (!paulis) {
    fprintf(stderr, "Error: Memory allocation failed.\n");
    return;
  }

  int pass = 1;
  for (uint64_t i = 0; i < total_strings; i++) {
    uint64_t value = i;
    for (unsigned int j = 0; j < N; j++) {
      paulis[N - j - 1] = pauli_chars[value % 4];
      value /= 4;
    }

    PauliStringC *pString =
        pauli_string_init_as_chars_c(N, 1.0 + 0.0 * (double complex)I, paulis);
    if (!pString) {
      fprintf(stderr, "Error: Failed to create Pauli string.\n");
      continue;
    }

    if (pString->encoding != i) {
      printf("Incorrect encoding at %llu != %llu!\n", pString->encoding, i);
      char *p_str = pauli_string_to_string_c(pString);
      printf("%s\n", p_str);
      free(p_str);
      pass = 0;
    }

    free_pauli_string_c(pString);
    if (!pass) {
      break;
    }
  }
  free(paulis);
  if (pass) {
    printf("Pauli encoding test completed successfully for %u qubits.\n", N);
  }
}

void test_pauli_sum_get_pauli_strings(void) {
  unsigned int paulis0[] = {3, 2, 0, 1};
  unsigned int paulis2[] = {2, 2, 0, 1};
  char paulis1[] = {'I', 'X', 'Y', 'I'};
  char paulis3[] = {'I', 'X', 'I', 'I'};

  char *pSum0_str;

  PauliStringC *pString0;
  PauliStringC *pString1;
  PauliStringC *pString2;
  PauliStringC *pString3;

  PauliSumC *pSum0;

  pString0 = pauli_string_init_as_ints_c(4, (double complex)1, paulis0);
  pString1 = pauli_string_init_as_chars_c(4, (double complex)I, paulis1);
  pString2 = pauli_string_init_as_ints_c(4, (double complex)1, paulis2);
  pString3 = pauli_string_init_as_chars_c(4, (double complex) - I, paulis3);

  pSum0 = pauli_sum_init_c(4);
  pauli_sum_append_pauli_string_c(pSum0, pString0);
  pauli_sum_append_pauli_string_c(pSum0, pString1);
  pauli_sum_append_pauli_string_c(pSum0, pString2);
  pauli_sum_append_pauli_string_c(pSum0, pString3);

  PauliStringC **pauli_strings = get_pauli_strings_c(pSum0);
  if (!pauli_strings) {
    fprintf(stderr, "Error: Failed to retrieve Pauli strings.\n");
    return;
  }

  pSum0_str = pauli_sum_to_string_c(pSum0);
  printf("%s contains %u Pauli strings:\n", pSum0_str, pSum0->p);
  free(pSum0_str);

  for (unsigned int i = 0; i < pSum0->p; i++) {
    char *p_str = pauli_string_to_string_c(pauli_strings[i]);
    printf("Pauli string %u: %s\n", i, p_str);
    free(p_str);
  }

  free(pauli_strings);
  free_pauli_sum_c(pSum0);
}

int main(void) {
  printf("\nTesting PauliString initializaiton, scaling, and freeing\n");
  test_pauli_string_initialization_scaling_freeing();
  printf("\nTesting PauliString Multiplication\n");
  test_pauli_string_multiplication();
  printf("\nTesting PauliSum initialization, scaling, and freeing\n");
  test_pauli_sum_initialization_scaling_freeing();
  printf("\nTesting PauliSum Adjoint\n");
  test_pauli_sum_adjoint();
  printf("\nTesting PauliSum Multiplication\n");
  test_pauli_sum_multiplication();
  printf("\nTesting PauliSum Addition\n");
  test_pauli_sum_addition();
  // printf("\nTesting Pauli String Encoding\n");
  // test_pauli_encoding(5);
  printf("\nTesting Pauli Sum Retrieval\n");
  test_pauli_sum_get_pauli_strings();
}
